atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/CandidatePagingState.java

```diff
@@ -0,0 +1,180 @@
+                // can cause data corruption.
+                Preconditions.checkArgument(cellTs.ts > maxSeenTimestampForCurrentCell,
+                        "Timestamps for each cell must be fed in strictly increasing order");
+            }
+            maxSeenTimestampForCurrentCell = cellTs.ts;
+            currentIsLatestValueEmpty = cellTs.isEmptyValue;
+            currentCellTimestamps.add(cellTs.ts);
+            cellTsPairsExamined += 1;
+        }
+        if (cellTsPairs.size() < maxCellTsPairsExpected) {
+            getCurrentCandidate().ifPresent(candidates::add);
+            reachedEnd = true;
+        }
+        return new BatchResult(candidates, reachedEnd);
+    }
+
+    public Optional<StartingPosition> getNextStartingPosition() {
+        if (reachedEnd) {
+            return Optional.empty();
+        } else {
+            return Optional.of(new StartingPosition(currentRowName, currentColName, getNextStartTimestamp()));
+        }
+    }
+
+    public void restartFromNextRow() {
+        @Nullable byte[] nextRow = RangeRequests.getNextStartRowUnlessTerminal(false, currentRowName);
+        if (nextRow == null) {
+            reachedEnd = true;
+        } else {
+            currentRowName = nextRow;
+            currentColName = PtBytes.EMPTY_BYTE_ARRAY;
+            maxSeenTimestampForCurrentCell = null;
+            currentCellTimestamps.clear();
+            currentIsLatestValueEmpty = false;
+            reachedEnd = false;
+        }
+    }
+
+    private Optional<CandidateCellForSweeping> getCurrentCandidate() {
+        if (currentCellTimestamps.isEmpty()) {
+            return Optional.empty();
+        } else {
+            return Optional.of(ImmutableCandidateCellForSweeping.builder()
+                    .cell(Cell.create(currentRowName, currentColName))
+                    .sortedTimestamps(currentCellTimestamps.toArray())
+                    .isLatestValueEmpty(currentIsLatestValueEmpty)
+                    .numCellsTsPairsExamined(cellTsPairsExamined)
+                    .build());
+        }
+    }
+
+    @Nullable
+    private Long getNextStartTimestamp() {
+        if (maxSeenTimestampForCurrentCell == null) {
+            return null;
+        } else {
+            // This can never happen because we request 'WHERE ts < ?', where '?' is some 'long' value.
+            // So if the timestamp is strictly less than another 'long', it can not be Long.MAX_VALUE.
+            // But we check anyway for general paranoia reasons.
+            Preconditions.checkState(maxSeenTimestampForCurrentCell != Long.MAX_VALUE,
+                    "Timestamps must be strictly less than Long.MAX_VALUE");
+            return maxSeenTimestampForCurrentCell + 1;
+        }
+    }
+
+    private Optional<CandidateCellForSweeping> checkCurrentCellAndUpdateIfNecessary(KvsEntryInfo cellTs) {
+        if (!isCurrentCell(cellTs.rowName, cellTs.colName)) {
```

Comment:
This looks like an ideal place to reduce branching, since inverting this check would _remove_ a `!`.